File: src/formats/mp3.lua
-- MP3 playback through dr_mp3, which build.lua compiles into the native library.
local ffi = require("ffi")
local native = require("treble.native")
local Tags = require("treble.tags")

ffi.cdef([[
	void *treble_mp3_open_file(const char *path);
	void *treble_mp3_open_memory(const void *data, size_t size);
	unsigned int treble_mp3_channels(const void *handle);
	unsigned int treble_mp3_sample_rate(const void *handle);
	unsigned long long treble_mp3_frame_count(const void *handle);
	unsigned long long treble_mp3_read_s16(void *handle, short *out, unsigned long long frames);
	unsigned int treble_mp3_seek(void *handle, unsigned long long frame);
	void treble_mp3_close(void *handle);
	unsigned int treble_mp3_tag_kind(const void *handle);
	const unsigned char *treble_mp3_tag_data(const void *handle);
	unsigned long long treble_mp3_tag_size(const void *handle);
]])

---@class treble.formats.Mp3Source: treble.Source
---@field private handle ffi.cdata*
---@field tags treble.Tags?
local Mp3Source = {}
Mp3Source.__index = Mp3Source

---@param handle ffi.cdata*
---@return treble.formats.Mp3Source
function Mp3Source.new(handle)
	local lib = native.lib
	local kind = lib.treble_mp3_tag_kind(handle)

	---@type treble.Tags?
	local tags = nil
	if kind ~= 0 then
		local size = tonumber(lib.treble_mp3_tag_size(handle))
		local raw = ffi.string(lib.treble_mp3_tag_data(handle), size)

		tags = kind == 2 and Tags.fromId3v2(raw) or Tags.fromId3v1(raw)
	end

	return setmetatable({
		handle = handle,
		tags = tags,
		channels = lib.treble_mp3_channels(handle),
		sampleRate = lib.treble_mp3_sample_rate(handle),
		frameCount = tonumber(lib.treble_mp3_frame_count(handle)),
	}, Mp3Source)
end

---@param out ffi.cdata*
---@param frames number
function Mp3Source:read(out, frames)
	return tonumber(native.lib.treble_mp3_read_s16(self.handle, out, frames))
end

---@param frame number
function Mp3Source:seek(frame)
	return native.lib.treble_mp3_seek(self.handle, frame) == 1
end

function Mp3Source:close()
	if self.handle ~= nil then
		native.lib.treble_mp3_close(self.handle)
		self.handle = nil
	end
end

local Mp3 = {}

---@param path string
---@return treble.formats.Mp3Source? source
---@return string? err
function Mp3.fromPath(path)
	local handle = native.lib.treble_mp3_open_file(path)
	if handle == nil then
		return nil, "Failed to open MP3 file: " .. path
	end

	local source = Mp3Source.new(handle)
	return source
end

--- Reads from memory that must stay alive until the source is closed.
---@param pointer ffi.cdata*
---@param size number
---@return treble.formats.Mp3Source? source
---@return string? err
function Mp3.fromMemory(pointer, size)
	local handle = native.lib.treble_mp3_open_memory(pointer, size)
	if handle == nil then
		return nil, "Failed to decode MP3 data"
	end

	local source = Mp3Source.new(handle)
	return source
end

--- Reports whether the content starts with an MP3 frame or an ID3 tag.
---@param content string
function Mp3.isValid(content)
	if string.sub(content, 1, 3) == "ID3" then
		return true
	end

	local first, second = string.byte(content, 1, 2)

	return first == 0xFF and second ~= nil and second >= 0xE0
end

return Mp3

File: src/native/mp3.c
/* The C surface treble keeps for MP3 decoding.
 *
 * dr_mp3 stores its decoder state in a struct whose size only its own header
 * knows, and it exposes the sample rate and channel count as struct fields
 * rather than through functions. LuaJIT cannot see either, so this shim owns the
 * struct and hands Lua an opaque handle, plus the two accessors dr_mp3 lacks.
 */

#define DR_MP3_IMPLEMENTATION
#include "dr_mp3.h"

#include <stdlib.h>
#include <string.h>

typedef struct treble_mp3 {
	drmp3 decoder;
	drmp3_uint64 cursor;
	/* Tags are copied out of the decoder: the pointers dr_mp3 hands the callback
	 * do not outlive the open call. */
	unsigned char* tagData;
	size_t tagSize;
	unsigned int tagKind;
} treble_mp3;

/* 0 for none, 1 for ID3v1, 2 for ID3v2. */
static void treble_mp3_capture(treble_mp3* handle, const drmp3_metadata* metadata)
{
	unsigned int kind = metadata->type == DRMP3_METADATA_TYPE_ID3V2 ? 2 : 1;

	/* An ID3v2 tag wins over the fixed size v1 footer. */
	if (handle->tagData != NULL && kind == 1) {
		return;
	}

	unsigned char* copy = malloc(metadata->rawDataSize);
	if (copy == NULL) {
		return;
	}

	memcpy(copy, metadata->pRawData, metadata->rawDataSize);
	free(handle->tagData);

	handle->tagData = copy;
	handle->tagSize = metadata->rawDataSize;
	handle->tagKind = kind;
}

static void treble_mp3_on_meta(void* userData, const drmp3_metadata* metadata)
{
	treble_mp3_capture((treble_mp3*)userData, metadata);
}

void* treble_mp3_open_file(const char* path)
{
	treble_mp3* handle = calloc(1, sizeof(treble_mp3));
	if (handle == NULL) {
		return NULL;
	}

	if (drmp3_init_file_with_metadata(&handle->decoder, path, treble_mp3_on_meta, handle, NULL) == DRMP3_FALSE) {
		free(handle);
		return NULL;
	}

	return handle;
}

/* The data pointer must stay valid and unchanged for the life of the handle. */
void* treble_mp3_open_memory(const void* data, size_t size)
{
	treble_mp3* handle = calloc(1, sizeof(treble_mp3));
	if (handle == NULL) {
		return NULL;
	}

	if (drmp3_init_memory_with_metadata(&handle->decoder, data, size, treble_mp3_on_meta, handle, NULL) == DRMP3_FALSE) {
		free(handle);
		return NULL;
	}

	return handle;
}

unsigned int treble_mp3_channels(const void* handle)
{
	return ((const treble_mp3*)handle)->decoder.channels;
}

unsigned int treble_mp3_sample_rate(const void* handle)
{
	return ((const treble_mp3*)handle)->decoder.sampleRate;
}

/* The frame count when the file states it, and 0 when it does not.
 *
 * drmp3_get_pcm_frame_count falls back to walking every frame of a file whose
 * Xing/Info tag is missing, which is fine on a local disk and unacceptable on a
 * network mount. A source that cannot state its length reports 0 instead, and the
 * player treats the end of the data as the end of the track.
 */
unsigned long long treble_mp3_frame_count(const void* handle)
{
	const drmp3* decoder = &((const treble_mp3*)handle)->decoder;

	if (decoder->totalPCMFrameCount == DRMP3_UINT64_MAX) {
		return 0;
	}

	return drmp3_get_pcm_frame_count((drmp3*)decoder);
}

unsigned long long treble_mp3_cursor(const void* handle)
{
	return ((const treble_mp3*)handle)->cursor;
}

/* Returns the frames written to out, which is frames * channels samples wide. */
unsigned long long treble_mp3_read_s16(void* handle, short* out, unsigned long long frames)
{
	treble_mp3* mp3 = handle;
	drmp3_uint64 read = drmp3_read_pcm_frames_s16(&mp3->decoder, frames, out);
	mp3->cursor += read;

	return read;
}

unsigned int treble_mp3_seek(void* handle, unsigned long long frame)
{
	treble_mp3* mp3 = handle;
	if (drmp3_seek_to_pcm_frame(&mp3->decoder, frame) == DRMP3_FALSE) {
		return 0;
	}

	mp3->cursor = frame;

	return 1;
}

/* 0 for none, 1 for ID3v1, 2 for ID3v2. */
unsigned int treble_mp3_tag_kind(const void* handle)
{
	return ((const treble_mp3*)handle)->tagKind;
}

const unsigned char* treble_mp3_tag_data(const void* handle)
{
	return ((const treble_mp3*)handle)->tagData;
}

unsigned long long treble_mp3_tag_size(const void* handle)
{
	return ((const treble_mp3*)handle)->tagSize;
}

void treble_mp3_close(void* handle)
{
	treble_mp3* mp3 = handle;
	if (mp3 == NULL) {
		return;
	}

	drmp3_uninit(&mp3->decoder);
	free(mp3->tagData);
	free(mp3);
}
